markets/storageadapter/getcurrentdealinfo.go

```diff
@@ -0,0 +1,59 @@
+package storageadapter
+
+import (
+	"bytes"
+	"context"
+
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/filecoin-project/go-state-types/exitcode"
+	"github.com/filecoin-project/lotus/api"
+	"github.com/filecoin-project/lotus/chain/actors/builtin/market"
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/ipfs/go-cid"
+	"golang.org/x/xerrors"
+)
+
+type getCurrentDealInfoAPI interface {
+	StateMarketStorageDeal(context.Context, abi.DealID, types.TipSetKey) (*api.MarketDeal, error)
+	StateSearchMsg(context.Context, cid.Cid) (*api.MsgLookup, error)
+}
+
+// GetCurrentDealInfo gets current information on a deal, and corrects the deal ID as needed
+func GetCurrentDealInfo(ctx context.Context, ts *types.TipSet, api getCurrentDealInfoAPI, dealID abi.DealID, publishCid *cid.Cid) (abi.DealID, *api.MarketDeal, error) {
+	marketDeal, dealErr := api.StateMarketStorageDeal(ctx, dealID, ts.Key())
+	if dealErr == nil {
+		return dealID, marketDeal, nil
+	}
+	if publishCid == nil {
+		return dealID, nil, dealErr
+	}
+	// attempt deal id correction
+	lookup, err := api.StateSearchMsg(ctx, *publishCid)
```

Comment:
This can be slow, especially when the message is 100-1000s of epochs old (~100ms-10s).

We should try to match marketDeal above to see if it's what we expect, and not do this search when it does match

Reply:
So check my logic on this but:
1. If deal id has not changed, we should never get to this line (cause of line 24)
2. If it has we will execute this slow search, and then we will return the new deal id, this will overwrite the existing deal id on lines 24 & 50 of ondealsectorcommitted.go. So the next time the function is called it will be called with the correct deal id, and hopefully therefore we will not make it past line 24.
3. As such, the only time this slow search is called is when the chain actually reorgs and the deal id changes?

Reply:
The only call to this function is line 74 of ondealsectorcommitted.go -- and this should be called exactly once.

Reply:
One thing I wanted to ask about re: testing the old deal ID first. If the deal ID has changed, presumably the old deal ID will not appear in the list of deal IDs in the precommit, and therefore testing for a match won't work? Actually, I wasn't sure about this -- is it possible a valid precommit message can contain an outdated dealID that isn't in state any longer? my reasoning for calling this function first is that presumably if the deal ID changed, it won't appear in the list of dealIDs for the precommit.

Reply:
deal IDs are sequential, so it's not impossible that a precommit will contain a deal with the same old dealID for a different deal

Reply:
@magik6k so we need to make sure to look up the new deal id as needed before we look at pre-commit.

This raises a seperate question -- is the check on line 24 of getcurrentdealinfo.go then valid? I.e. do we need to do something more than verify the state returns without error? Should we have a proposal CID and verify it matches the proposal part of the state?